File: frontend/src/com/s2c.ts
import { Mapper } from '../util'
import { ProjectInfo, ProjectId, FileInfo, FileId, Doc } from './types'

export type Lock = Mapper<{
  Unlock: {}
  LockBy: {}
  LockByMe: {}
}>

export type Server2Client = Mapper<{
  Projects: {
    list: ProjectInfo[]
  }
  Project: {
    id: ProjectId
    msg: Server2ClientProject
  }
}>

export type Server2ClientProject = Mapper<{
  Files: {
    list: FileInfo[]
  }
  UpdateInfo: {
    info: ProjectInfo
  }
  File: {
    id: FileId
    msg: Server2ClientProjectFile
  }
}>

export type Server2ClientProjectFile = Mapper<{
  FileLock: { lock: Lock }
  FileSource: { src: string }
  FileDoc: { doc: Doc }
}>

File: src/s2c.rs
use actix::Message;

use serde::Serialize;

use crate::project::file::{Doc, FileId, FileInfo};
use crate::project::{ProjectId, ProjectInfo};

#[derive(Serialize, Clone)]
#[serde(tag = "type")]
pub enum Lock {
    Unlock,
    LockBy {},
    LockByMe,
}

#[derive(Serialize, Clone, Message)]
#[serde(tag = "type")]
pub enum Server2Client {
    Projects {
        list: Vec<ProjectInfo>,
    },
    Project {
        id: ProjectId,
        msg: Server2ClientProject,
    },
}

#[derive(Serialize, Clone)]
#[serde(tag = "type")]
pub enum Server2ClientProject {
    Files {
        list: Vec<FileInfo>,
    },
    UpdateInfo {
        info: ProjectInfo,
    },
    File {
        id: FileId,
        msg: Server2ClientProjectFile,
    },
}

#[derive(Serialize, Clone)]
#[serde(tag = "type")]
pub enum Server2ClientProjectFile {
    FileLock { lock: Lock },
    FileSource { src: String },
    FileDoc { doc: Doc },
}
